runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+use sr_primitives::transaction_validity::InvalidTransaction;
+
+/// Parachain registration API.
+pub trait Registrar<AccountId> {
+	/// Create a new unique parachain identity for later registration.
+	fn new_id() -> ParaId;
+
+	/// Register a parachain with given `code` and `initial_head_data`. `id` must not yet be registered or it will
+	/// result in a error.
+	fn register_para(
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
+		initial_head_data: Vec<u8>
+	) -> Result;
+
+	/// Deregister a parachain with given `id`. If `id` is not currently registered, an error is returned.
+	fn deregister_para(id: ParaId) -> Result;
+}
+
+impl<T: Trait> Registrar<T::AccountId> for Module<T> {
+	fn new_id() -> ParaId {
+		<NextFreeId>::mutate(|n| { let r = *n; *n = ParaId::from(u32::from(*n) + 1); r })
+	}
+
+	fn register_para(
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
+		initial_head_data: Vec<u8>
+	) -> Result {
+		ensure!(!Paras::exists(id), "Parachain already exists");
+		if let Scheduling::Always = info.scheduling {
+			Parachains::mutate(|parachains|
+				match parachains.binary_search(&id) {
+					Ok(_) => Err("Parachain already exists"),
+					Err(idx) => {
+						parachains.insert(idx, id);
+						Ok(())
+					}
+				}
+			)?;
+		}
+		<parachains::Module<T>>::initialize_para(id, code, initial_head_data);
+		Paras::insert(id, info);
+		Ok(())
+	}
+
+	fn deregister_para(id: ParaId) -> Result {
+		let info = Paras::take(id).ok_or("Invalid id")?;
+		if let Scheduling::Always = info.scheduling {
+			Parachains::mutate(|parachains|
+				parachains.binary_search(&id)
+					.map(|index| parachains.remove(index))
+					.map_err(|_| "Invalid id")
+			)?;
+		}
+		<parachains::Module<T>>::cleanup_para(id);
+		Paras::remove(id);
+		Ok(())
+	}
+}
+
+type BalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+pub trait Trait: parachains::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// The aggregated origin type must support the parachains origin. We require that we can
+	/// infallibly convert between this origin and the system origin, but in reality, they're the
+	/// same type, we just can't express that to the Rust type system without writing a `where`
+	/// clause everywhere.
+	type Origin: From<<Self as system::Trait>::Origin>
+		+ Into<result::Result<parachains::Origin, <Self as Trait>::Origin>>;
+
+	/// The system's currency for parathread payment.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// The deposit to be paid to run a parathread.
+	type ParathreadDeposit: Get<BalanceOf<Self>>;
+
+	/// Handler for when two ParaIds are swapped.
+	type OnSwap: OnSwap;
+}
+
+/// The number of items in the parathread queue, aka the number of blocks in advance to schedule
+/// parachain execution.
+const QUEUE_SIZE: usize = 2;
+
+/// The number of rotations that you will have as grace if you miss a block.
+const MAX_RETRIES: u32 = 3;
```

Comment:
Any reason not to make these trait constants?

Reply:
Not really. Will do it.